design/splitfrr-proposal.md

```diff
@@ -0,0 +1,522 @@
+      - 192.168.11.0/32
+      neighbors:
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toAdvertise:
+          allowed:
+          - 192.168.11.0/32
+```
+
+The resulting configuration will be something like:
+
+```yaml
+  bgp:
+    routers:
+    - asn: 64512
+      prefixesIpV4:
+      - 192.168.10.0/32
+      - 192.168.11.0/32
+      neighbors:
+      - address: 172.18.0.5
+        asn: 64512
+        port: 179
+      - address: 172.18.0.6
+        asn: 64512
+        port: 179
+        toAdvertise:
+          allowed:
+          - 192.168.10.0/32
+          - 192.168.11.0/32
+        toReceive:
+          allowed:
+            allowAll: true
+```
+
+The FRR configuration rendering mechanism is going to be the same as the one used today by MetalLB.
+The idea is to reuse the same packages with a layer that translates this API to the
+internal one.
+
+Based on the API, the configuration will be rendered through a template, and the `reload-frr.py` script will
+be invoked via a signal (the very same mechanism we have in MetalLB).
+
+The structure of the frr configuration file is going to be something like:
+
+- a router
+- the prefixes for each ipfamily
+- a set of neighbors
+- route-rules for each neighbor
+- prefix lists associated to each route-rule
+
+### Exposing the status
+
+We should be able to expose:
+
+- The rendered frr configuration for each node
+- The failure of a given configuration rendering and the reason
+- Details about the status of BFD / BGP sessions from one node to each neighbor
+
+A prototype that covers the first two items might look like:
+
+```yaml
+type FRRNodeStatus struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Status FRRConfigurationStatus `json:"status,omitempty"`
+}
+
+type ConfigurationStatus struct {
+  DesiredConfiguration string
+  RunningConfiguration string
+  Progress Status
+}
+
+type Status struct {
+  Phase string
+  Message string
+}
+```
+
+For the item about the session status, it's basically covered in the [metallb status proposal](https://github.com/metallb/metallb/blob/main/design/crd-status.md#bgp--bfd-session-state).
+
+Additionally, we might consider either allowing MetalLB to probe directly the
+FRR instance or expose the same set of metrics that MetalLB is currently exposing.
+
+### Breaking the node
+
+Applying the wrong configuration / fetching the wrong routes might break the node.
+We should have a mechanism in place to avoid getting routes that overlap the pods / service CIDRs and any other sensible route that a cluster might require.
+
+Moreover, we should explore the possibility of having sanity checks for the node and suspend the FRR configuration in that scenario.
+
+### Deployment
+
+This new FRR daemon will live in a separate repo, and ideally it's going to
+be deployable on its own.
+
+On the other hand, the MetalLB Operator will be changed so it can deploy this
+new component too.
```

Comment:
as a starter I'd let the operator deploy frr together with metallb, because metallb requires it to function. IIRC, olm allows to specify operator dependencies or something like that, so it might be that in the future this thing has its own operator.